wire-ios-sync-engine/Source/Synchronization/EventProcessor.swift

```diff
@@ -24,139 +24,111 @@ extension NSNotification.Name {
         eventProcessingTracker: EventProcessingTrackerProtocol,
-        earService: EARServiceInterface
+        earService: EARServiceInterface,
+        eventConsumers: [ZMEventConsumer]
     ) {
         self.syncContext = storeProvider.syncContext
         self.eventContext = storeProvider.eventContext
-        self.syncStatus = syncStatus
         self.eventDecoder = EventDecoder(eventMOC: eventContext, syncMOC: syncContext)
         self.eventProcessingTracker = eventProcessingTracker
         self.earService = earService
-        self.eventBuffer = ZMUpdateEventsBuffer(updateEventProcessor: self)
+        self.bufferedEvents = []
+        self.eventConsumers = eventConsumers
     }
 
     // MARK: Methods
 
-    /// Process previously received events if we are ready to process events.
-    ///
-    /// - Returns: **True** if there are still more events to process
-    @objc
-    public func processEventsIfReady() -> Bool { // TODO jacob shouldn't be public
-        WireLogger.updateEvent.info("process events if ready")
-
-        guard isReadyToProcessEvents else {
-            WireLogger.updateEvent.info("not ready to process events")
-            return true
-        }
-
-        eventBuffer?.processAllEventsInBuffer()
+    func bufferEvents(_ events: [ZMUpdateEvent]) async {
+        guard !DeveloperFlag.ignoreIncomingEvents.isOn else { return }
+        bufferedEvents.append(contentsOf: events)
+    }
 
-        var hasMoreEventsToProcess = false
-        do {
-            try processEvents(callEventsOnly: false)
-        } catch {
-            hasMoreEventsToProcess = true
+    func processEvents(_ events: [ZMUpdateEvent]) async throws {
+        try await enqueueTask {
+            guard !DeveloperFlag.ignoreIncomingEvents.isOn else { return }
+            let publicKeys = try? self.earService.fetchPublicKeys()
+            let decryptedEvents = await self.eventDecoder.decryptAndStoreEvents(events, publicKeys: publicKeys)
+            await self.processBackgroundEvents(decryptedEvents)
+            await self.requestToCalculateBadgeCount()
+            let isLocked = await self.syncContext.perform { self.syncContext.isLocked }
+            try? await self.processEvents(callEventsOnly: isLocked)
         }
+    }
 
-        return hasMoreEventsToProcess
+    func processBufferedEvents() async throws {
+        let events = bufferedEvents
+        bufferedEvents.removeAll()
+        try await processEvents(events)
     }
 
-    func processPendingCallEvents() throws {
-        try syncContext.performGroupedAndWait { _ in
-            self.eventBuffer?.processAllEventsInBuffer()
-            try self.processEvents(callEventsOnly: true)
+    private func enqueueTask(_ block: @escaping @Sendable () async throws -> Void) async throws {
+        processingTask = Task { [processingTask] in
+            _ = await processingTask?.result
+            return try await block()
         }
+        _ = await processingTask?.result
```

Comment:
@typfel looking at this again, I noticed we don't throw the error so the errors are not propagated. 

is there a good reason for it?